sql/wsrep_mysqld.h

```diff
@@ -369,7 +370,118 @@ class Log_event;
 int wsrep_ignored_error_code(Log_event* ev, int error);
 int wsrep_must_ignore_error(THD* thd);
 
-bool wsrep_replicate_GTID(THD* thd);
+struct wsrep_server_gtid_t
+{
+  uint32 domain_id;
+  uint32 server_id;
+  uint64 seqno;
+};
+class Wsrep_gtid_server
+{
+private:
+  struct gtid_wait_t
+  {
+    uint64 seqno;
+    mysql_cond_t cond;
+  };
+public:
+  uint32 domain_id;
+  uint32 server_id;
+  void cleanup_wait_list()
+  {
+    signal_wait_list(0, true);
+  }
+  void gtid(const wsrep_server_gtid_t& gtid)
+  {
+    domain_id=  gtid.domain_id;
+    server_id=  gtid.server_id;
+    m_seqno=    gtid.seqno;
+  }
+  wsrep_server_gtid_t gtid()
+  {
+    wsrep_server_gtid_t gtid;
+    gtid.domain_id= domain_id;
+    gtid.server_id= server_id;
+    gtid.seqno=     m_seqno;
+    return gtid;
+  }
+  void seqno(const uint64 seqno) { m_seqno= seqno; }
+  uint64 seqno_last_seen() const  {  return m_seqno; }
+  uint64 seqno_last_written() const { return m_last_local_generated_seqno; }
+  uint64 seqno_inc(bool local) 
+  {
+    m_seqno++;
+    uint64 seqno= m_seqno;
+    if (local) m_last_local_generated_seqno= seqno;
+    signal_wait_list(m_seqno, false);
+    return m_seqno;
+  }
+  const wsrep_server_gtid_t& undefined()
+  {
+    return m_undefined;
+  }
+  int wait_gtid_upto(const uint64_t& seqno, uint timeout)
```

Comment:
`seqno` should be passed by value. As a rule of thumb, always pass by value if the size of the function argument type is known and is 8 bytes or less.